core/drivers/crypto/hisilicon/sec_cipher.c

```diff
@@ -0,0 +1,771 @@
+		goto free_c_ctx;
+	}
+
+	if (c_ctx->qp->qm->version == HISI_QM_HW_V2) {
+		c_ctx->qp->fill_sqe = sec_cipher_bd_fill;
+		c_ctx->qp->parse_sqe = sec_cipher_bd_parse;
+	} else {
+		c_ctx->qp->fill_sqe = sec_cipher_bd3_fill;
+		c_ctx->qp->parse_sqe = sec_cipher_bd3_parse;
+	}
+
+	c_ctx->offs = 0;
+	*ctx = c_ctx;
+
+	return TEE_SUCCESS;
+
+free_c_ctx:
+	free(c_ctx);
+
+	return ret;
+}
+
+static void sec_cipher_ctx_free(void *ctx)
+{
+	struct sec_cipher_ctx *c_ctx = ctx;
+
+	if (!c_ctx)
+		return;
+
+	hisi_qm_release_qp(c_ctx->qp);
+	memzero_explicit(c_ctx->key, c_ctx->key_len);
+	free(c_ctx);
+}
+
+static TEE_Result sec_cipher_initialize(struct drvcrypt_cipher_init *dinit)
+{
+	struct sec_cipher_ctx *c_ctx = NULL;
+	TEE_Result ret = TEE_SUCCESS;
+
+	if (!dinit || !dinit->ctx || !dinit->key1.data) {
+		EMSG("drvcrypt_cipher init param error");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	c_ctx = dinit->ctx;
+
+	ret = sec_cipher_set_key(c_ctx, dinit->key1.data, dinit->key1.length,
+				 dinit->key2.data, dinit->key2.length);
+	if (ret)
+		return ret;
+
+	ret = sec_cipher_set_iv(c_ctx, dinit->iv.data, dinit->iv.length);
+	if (ret)
+		return ret;
+
+	c_ctx->encrypt = dinit->encrypt;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result sec_cipher_cryptlen_check(struct sec_cipher_ctx *c_ctx,
+					    size_t length)
+{
+	if (c_ctx->mode == C_MODE_XTS && length < AES_SM4_BLOCK_SIZE) {
+		EMSG("Invalid xts src len");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if ((c_ctx->mode == C_MODE_ECB || c_ctx->mode == C_MODE_CBC) &&
+	    (length & (AES_SM4_BLOCK_SIZE - 1))) {
+		EMSG("Invalid ecb cbc src len");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result sec_cipher_param_check(struct drvcrypt_cipher_update *dupdate)
+{
+	struct sec_cipher_ctx *c_ctx = NULL;
+
+	if (!dupdate || !dupdate->src.data || !dupdate->dst.data ||
+	    dupdate->src.length != dupdate->dst.length ||
+	    dupdate->src.length > MAX_CIPHER_LENGTH || !dupdate->src.length) {
+		EMSG("Dupdate input param error");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	c_ctx = dupdate->ctx;
+	switch (c_ctx->alg) {
+	case C_ALG_SM4:
+	case C_ALG_AES:
+		if (sec_cipher_cryptlen_check(c_ctx, dupdate->src.length))
+			return TEE_ERROR_BAD_PARAMETERS;
+		break;
+	case C_ALG_DES:
+	case C_ALG_3DES:
+		if (dupdate->src.length % TEE_DES_BLOCK_SIZE) {
+			EMSG("Invalid src len");
+			return TEE_ERROR_BAD_PARAMETERS;
+		}
+		break;
+	default:
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result sec_alloc_buffer(struct sec_cipher_ctx *c_ctx)
+{
+	c_ctx->in = malloc(c_ctx->len);
+	if (!c_ctx->in) {
+		EMSG("Fail to alloc c_in buf");
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	c_ctx->in_dma = virt_to_phys(c_ctx->in);
+
+	c_ctx->out = malloc(c_ctx->len);
+	if (!c_ctx->out) {
+		EMSG("Fail to alloc c_out buf");
+		goto free_c_in;
+	}
+
+	c_ctx->out_dma = virt_to_phys(c_ctx->out);
+
+	return TEE_SUCCESS;
+
+free_c_in:
+	free(c_ctx->in);
```

Comment:
This leaves `c_ctx->in` with an invalid non-NULL reference. I think it would be better to set `c_ctx->in = NULL` here as `.free_ctx` handler can be called after `sec_cipher_update()` failure.

Reply:
done